### PR TITLE
🪟🚨 Refactor connector form code

### LINKED ISSUE #14250: (Resolve) Tech debt of the ConnectorForm
Brainstomring issue about problem with the ServiceForm component. Please feel free to list all issues that should be addressed to make this component more stable and easier to maintain:

- [x] https://github.com/airbytehq/airbyte/issues/13056
- [x] Only render connector form component if the selected connector definition specification is loaded (move loading state and spinner into ConnectorCard) https://github.com/airbytehq/airbyte/pull/20127
- [x] Remove component override for name field and make it a normal field in the schema with title and description https://github.com/airbytehq/airbyte/pull/20132
- [x] Move unfinished flows into local state in ConfirmationControls (reset from the outside if formik state becomes non-dirty) https://github.com/airbytehq/airbyte/pull/20135
- [ ] Remove uiWidgetsInfo state by finding the enum/const field defining the selected oneOf and adding it to the condition form block to check the selected value by looking at the formik state
- [ ] Only do default value calculation once before rendering the formik component the first time (remove `PatchInitialValuesWithWidgetConfig` hack)
- [ ] Build yup schema once by flattening the oneOf conditions into a single object and adding `when` conditions on the found condition key (remove `RevalidateOnValidationSchemaChange` hack)
- [ ] Add https://github.com/airbytehq/airbyte/issues/17352 to default value calculation

* ~Pull out auth field hiding from `FormSection` to a higher level when generating from the spec which fields should be rendered.~
* ~`values.connectionConfiguration` plus flattened values are confusing (for a flattened form)~
* ~Rethink `default` behavior (should it be sending the value if the field is empty?), since it's working different depending on your field type atm~ -> extra issue

### LINKED ISSUE #15108: Don't change validationSchema based on input
Currently our code does change the `validationSchema` of the ServiceForm based on some values, e.g. if the user selects another oneOf value with a sub object the whole validationSchema will change to reflect that.

I think we should try to generate the `validationSchema` once for the connectors spec and not change it based on the form values. Yup supports [`when`](https://github.com/jquense/yup#schemawhenkeys-string--string-builder-object--values-any-schema--schema-schema) to actually annotate that specific parts of a schema are only in effect when specific criterias are met. I think we should rather use that functionality to generate one validationSchema, that uses `when` to validate the different choices of a `oneOf` JSONSchema.

Especially since Formik doesn't revalidate the form when the validationSchema changes, we need to have special hacks for that in place.

This would require https://github.com/airbytehq/airbyte/issues/13056 to be solved, so that the formik form actually only contain the individual connectors form, and thus the validationSchema could be kept unchanged while filling the form.

### PR DESCRIPTION
**Pull request by flash1293**
Closes https://github.com/airbytehq/airbyte/issues/14250
Closes https://github.com/airbytehq/airbyte/issues/15108

## What

This PR takes care of the rest of the connector form tech debt parts:
*  Remove uiWidgetsInfo state by finding the enum/const field defining the selected oneOf and adding it to the condition form block to check the selected value by looking at the formik state
* Only do default value calculation once before rendering the formik component the first time (remove PatchInitialValuesWithWidgetConfig hack)
* Build yup schema once by flattening the oneOf conditions into a single object and adding when conditions on the found condition key (remove RevalidateOnValidationSchemaChange hack)

## How

Selected conditions of a oneOf are not tracked via `uiWidgetsInfo` state anymore. Instead, the `FormBlock` data structure contains a new property "selectionPath" which is pointing to the shared const prop of all conditions. As this is part of the formik values, it can be looked up from there directly. A second new property "selectionConstValues" contains an array of possible values for the const field in the same order as the condition form blocks itself - checking which one is selected becomes a `selectionConstValues.indexOf(formikState.get(selectionPath))`.

The yup schema used for actual validation is not rebuilt based on the `uiWidgetsInfo` state anymore. Instead, the keys out of all conditions are flattened into a single yup object shape, with `when` conditions applying the correct sub-schema based on the value of the "selectionPath". If the selected condition does not specify a certain key at all, its schema is set to "strip" which removes it on casting.

Setting default values doesn't happen within a hack-component anymore but is part of the `useBuildForm` hook now which is also generating the form blocks data structure.

In a few places from within the new logic, errors are thrown if the rules of airbyte schema are violated. For these, there is a special error type which is caught by the error boundary and rendered with a nice message and a link to the documentation. This should not happen for any of the Airbyte-maintained connectors but might be a case for custom ones.

The uiWidgets terminology was used in inconsistent ways - this PR removes it all together - now this is the flow that's happening when rendering the form for the first time:
* json schema -> form blocks and initial default values
* json schema + form blocks -> yup schema

On update, only the formik values are updated.

As the initial form values on empty forms passed down from the connector card were re-built on every render, the init logic outlined above would sometimes happen multiple times (depending on how often react-query is re-rendering the component). By memoizing these initial values on the connector card level, it's only happening a single time.

## 🚨 User Impact 🚨

The snowflake destination in version `<=0.4.40` does not work together with the changes on this PR - existing connections will continue to work fine, but it's not possible to change the configuration. Please update the snowflake destination connector to `0.4.41`.

oneOf properties not following the rules described in [the documentation](https://docs.airbyte.com/connector-development/connector-specification-reference/#airbyte-modifications-to-jsonschema) will stop working in the UI - the form will crash and a meaningful error is shown which also links to the documentation:
Screenshot 2022-12-09 at 12 00 11

### CODE REVIEW

File: airbyte-webapp/src/views/Connector/ConnectorCard/ConnectorCard.tsx
Comment (flash1293) [reply airbyte-webapp/src/views/Connector/ConnectorCard/ConnectorCard.tsx:172]: Because props.connector only exists if isEditMode is truthy (it's a union type)

File: airbyte-webapp/src/views/Connector/ConnectorForm/useBuildForm.tsx
Comment (flash1293) [reply airbyte-webapp/src/views/Connector/ConnectorForm/useBuildForm.tsx:65]: As it's only called in here recursively this seemed cleaner to me (keeps it closer to the place where it's used). No strong opinion on these things though, let me know if you would prefer to move it.

File: airbyte-webapp/src/views/Connector/ConnectorCard/ConnectorCard.tsx
Comment (krishnaglick) [reply airbyte-webapp/src/views/Connector/ConnectorCard/ConnectorCard.tsx:172]: My only argument for doing the above method is if this hook changes the linter will pick up new deps.

File: airbyte-webapp/src/views/Connector/ConnectorForm/useBuildForm.tsx
Comment (krishnaglick) [reply airbyte-webapp/src/views/Connector/ConnectorForm/useBuildForm.tsx:65]: I'm partial to keeping unwrapped (non-useCallback, etc) methods at the top scope within the react render tree since every time this is called (which may be a small number of times) it's redefined.

Comment (flash1293) [reply airbyte-webapp/src/views/Connector/ConnectorForm/useBuildForm.tsx]: Wouldn't matter in this specific case, but I agree it's a good default. Improved

File: airbyte-webapp/src/views/Connector/ConnectorCard/ConnectorCard.tsx
Comment (flash1293) [reply airbyte-webapp/src/views/Connector/ConnectorCard/ConnectorCard.tsx]: Found a way to structure this better, thanks for the suggestion.

File: airbyte-webapp/src/core/form/schemaToFormBlock.ts
Comment (flash1293) [reply airbyte-webapp/src/core/form/schemaToFormBlock.ts]: Yes, otherwise typescript tries to be smart and assumes it will always be `null`:
Screenshot 2022-12-13 at 16 19 14

However, we can omit the type of the variable. Improved.

**Comment by flash1293** [comment]
Thanks for the review @krishnaglick , fixed the things you highlighted.

**Review by krishnaglick — APPROVED** [review]
Nothing else stands out to me but someone closer to these changes should definitely do a thorough review!

**Comment by flash1293** [comment]
> but someone closer to these changes should definitely do a thorough review!

Definitely, this is a change that can break quite a few things in subtle ways, let's make sure to test it thoroughly. I will give it another round myself.

File: airbyte-webapp/src/core/form/schemaToYup.ts
Comment (lmossman) [inline airbyte-webapp/src/core/form/schemaToYup.ts:59]: Why `yup.mixed()` instead of `yup.boolean()`?

File: airbyte-webapp/src/core/form/schemaToYup.ts
Comment (lmossman) [inline airbyte-webapp/src/core/form/schemaToYup.ts:43]: Why is the value type of the map `Array<[unknown, yup.AnySchema]>`? Shouldn't it be `Array<[string, yup.AnySchema]>`, since all properties are strings?

File: airbyte-webapp/src/core/form/schemaToYup.ts
Comment (lmossman) [inline airbyte-webapp/src/core/form/schemaToYup.ts:153]:
```suggestion
          throw new FormBuildError("mismatch between form fields and schema");
```

File: airbyte-webapp/src/core/form/schemaToYup.ts
Comment (lmossman) [inline airbyte-webapp/src/core/form/schemaToYup.ts:36]: This whole if block took me a few tries to understand as it is fairly complex. It would be great if we could simplify it a bit but I'm not immediately seeing a way to do that, and I don't think that should necessarily block merging

File: airbyte-webapp/src/core/form/schemaToYup.ts
Comment (lmossman) [inline airbyte-webapp/src/core/form/schemaToYup.ts:153]: This seems like more of an internal bug, so we should probably track this error in datadog so we can be aware of it when it happens. We have a service (app monitoring service) which can be used to track errors like this. Here is an example: https://github.com/airbytehq/airbyte/blob/2392acb845891f9c031b06feee9af2a2f975f1a7/airbyte-webapp/src/views/Connector/ConnectorForm/useAuthentication.tsx#L156-L159
Here is a slack post with more context about this service: https://airbytehq-team.slack.com/archives/C039SG3DQ8K/p1668506492916659

In fact, we may just want to track all FormBuildErrors?

**Review by lmossman — COMMENTED** [review]
Had several comments. Overall I think the changes make sense. They get a bit complex in some areas but that is likely just a result of this being a complex problem to solve.

I did not get a chance to test this but I will do so before I give it a full approval

File: airbyte-webapp/src/core/form/schemaToFormBlock.ts
Comment (flash1293) [reply airbyte-webapp/src/core/form/schemaToFormBlock.ts:44]: You are 100% right, fixed

File: airbyte-webapp/src/core/form/schemaToFormBlock.ts
Comment (flash1293) [reply airbyte-webapp/src/core/form/schemaToFormBlock.ts:41]: It would be, but if you do that, Typescript will auto-narrow down your type to just `null`:
Screenshot 2022-12-15 at 11 54 04

File: airbyte-webapp/src/core/form/schemaToFormBlock.ts
Comment (flash1293) [reply airbyte-webapp/src/core/form/schemaToFormBlock.ts:53]: The `Set` class is an ordered set in Javascript: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set

I know it's not important for the amount of data we are dealing with, but theoretically the intersect is slightly faster this way because the `has` method is constant time now instead of linear time for ``, reducing the algorithm from `n*n` to `n`.

That said, it's probably better to stick with arrays here to keep the logic simple as there's a lot of stuff going on already - rewrote it using arrays and the lodash intersection helper which I hope will do if fast (if not its easy to refactor later on)

File: airbyte-webapp/src/core/jsonSchema/index.ts
Comment (flash1293) [reply airbyte-webapp/src/core/jsonSchema/index.ts:13]: Good point, changed.

File: airbyte-webapp/src/core/form/schemaToYup.ts
Comment (flash1293) [reply airbyte-webapp/src/core/form/schemaToYup.ts:59]: I inherited this from the old code. It didn't strike me as exceptionally important as we render a switch anyway for these so the user can't really do wrong on them in the first place.

Still, changed it to keep the code unsurprising.

File: airbyte-webapp/src/core/form/schemaToYup.ts
Comment (flash1293) [reply airbyte-webapp/src/core/form/schemaToYup.ts:43]: The keys of the map correspond with the property keys in each of the possible `oneOf`s - those need to be strings. But the array of `Array<[unknown, yup.AnySchema]>` is a list of mappings from the value of the const property to the schema it implies - const properties can be anything, they are not limited to strings (in practice they are probably always strings, but numbers could be the case as well).

This is where the type for these comes from:
Screenshot 2022-12-15 at 12 13 15

`JSONSchema7Type` is pretty much everything you want:
Screenshot 2022-12-15 at 12 14 15

File: airbyte-webapp/src/core/form/schemaToYup.ts
Comment (flash1293) [reply airbyte-webapp/src/core/form/schemaToYup.ts:36]: Yeah, that's pretty dense code. I added an example to the comment to explain what's going on.

File: airbyte-webapp/src/core/form/schemaToYup.ts
Comment (flash1293) [reply airbyte-webapp/src/core/form/schemaToYup.ts:153]: Good idea, added that to the error boundary - if a form group error bubbles up, it is tracking the error using datadog.

**Comment by flash1293** [comment]
Thanks for the review, addressed all your points.

File: airbyte-webapp/src/core/form/schemaToYup.test.ts
Comment (lmossman) [inline airbyte-webapp/src/core/form/schemaToYup.test.ts:117]:
```suggestion
it("should build mixed schema for conditional case", () => {
```

File: airbyte-webapp/src/core/form/schemaToYup.test.ts
Comment (lmossman) [inline airbyte-webapp/src/core/form/schemaToYup.test.ts:128]: This doesn't make sense to me. Wouldn't we expect the `api_key` schema to be something like
```
api_key: yup
  .mixed()
  .when("type", { is: "api", then: (x) => x.string().matches(\\w{5}), otherwise: (x) => x })
  .when("type", { is: (val) => !["api"].includes(val), then: (x) => x.strip() })
```

same comment goes for the `redirect_uri` schema below, and the schemas at the end of this file

File: airbyte-webapp/src/core/form/schemaToYup.test.ts
Comment (lmossman) [inline airbyte-webapp/src/core/form/schemaToYup.test.ts:206]: This test implementation doesn't seem to line up with this description to me. I'm not sure what `topKey` and `topKey.subKey` are doing on lines 250-251. It seems like this should be nesting `credentials` under a few levels or something?

File: airbyte-webapp/src/core/form/schemaToYup.ts
Comment (lmossman) [reply airbyte-webapp/src/core/form/schemaToYup.ts:36]: The examples are super helpful, thanks!

File: airbyte-webapp/src/core/form/schemaToYup.ts
Comment (lmossman) [reply airbyte-webapp/src/core/form/schemaToYup.ts:53]: Would it work to make the type of `FormConditionItem.selectionConstValues` be something like `Array<JSONSchema7Type | undefined>` then?

Just asking because it feels like we should try to avoid `unknown` where we can and this feels like a place where we may be able to eliminate it

**Review by lmossman — COMMENTED** [review]
Had a few more comments, main ones being about how the tests look like they may be incorrect 

I also tested this locally with a few sources (postgres, linkedin ads, etc) and the behavior looked correct to me!

File: airbyte-webapp/src/core/form/schemaToFormBlock.ts
Comment (flash1293) [reply airbyte-webapp/src/core/form/schemaToFormBlock.ts:45]: I think there are two kinds of errors thrown from our code:
* Our fault - unexpected errors which are 100% on us because there is a bug in the code
* User fault - errors that can happen due to user input

I think the first kind doesn't need to be i18n-ified, it will most likely anyway not tell the user anything useful, it's just relevant for devs to figure out what's going on.

The second kind should be i18n-ified because ideally the user can fix it themselves.

I originally though these errors would be of the first kind, but on second thought they can also be seen as user errors, especially with the push to allow custom connectors. So I went ahead and translated them. `FormBuildError` is now only meant for errors which are caused by a json schema not following the documented rules. I changed one place where it was thrown back to `Error` because it can only be caused by a bug on our side.

File: airbyte-webapp/src/core/form/schemaToYup.test.ts
Comment (flash1293) [reply airbyte-webapp/src/core/form/schemaToYup.test.ts:128]: This test is only testing the structure of the yup schema which does not capture the actual conditions of the whens:
```json
 "api_key": {
              "type": "mixed",
              "_whitelist": {
                "list": {},
                "refs": {}
              },
              "_blacklist": {
                "list": {},
                "refs": {}
              },
              "exclusiveTests": {},
              "deps": [
                "type",
                "type"
              ],
              "conditions": [
                {
                  "refs": [
                    {
                      "key": "type",
                      "isContext": false,
                      "isValue": false,
                      "isSibling": true,
                      "path": "type"
                    }
                  ]
                },
                {
                  "refs": [
                    {
                      "key": "type",
                      "isContext": false,
                      "isValue": false,
                      "isSibling": true,
                      "path": "type"
                    }
                  ]
                }
              ],
```

The tests below this one (the stuff in `should schema build conditional case that validates correctly based on selection key`) test whether the `when`s actually kick in. 

Added a comment to explain better.

File: airbyte-webapp/src/core/form/schemaToYup.test.ts
Comment (flash1293) [reply airbyte-webapp/src/core/form/schemaToYup.test.ts:206]: I copied this test over and didn't adjust much about it. It is testing whether conditionals also work in nested cases, but I don't see how we need that test as the base schema is also nesting the oneOf in an object...

Removed the test as it's more confusing than helpful.

File: airbyte-webapp/src/core/form/schemaToYup.ts
Comment (flash1293) [reply airbyte-webapp/src/core/form/schemaToYup.ts:53]: I'm not too worried about using `unknown` as it's a clear marker that you shouldn't assume anything about this value -  it basically allows you to do nothing with it except for passing it around and checking whether it's the same value. 

Switching to `JSONSchema7Type` is closer to the truth though - updated the code (we can even drop the "undefined" as this would mean in one condition no `const` is set which is something we throw on).

**Comment by flash1293** [comment]
I clicked through all the sources and all the destinations and found one that doesn't work - the Snowflake destination.
It does not specify the "auth_type" property for the simple password case:
```json
[
    {
        "type": "object",
        "order": 0,
        "title": "OAuth2.0",
        "required": [
            "access_token",
            "refresh_token"
        ],
        "properties": {
            "auth_type": {
                "enum": [
                    "OAuth2.0"
                ],
                "type": "string",
                "const": "OAuth2.0",
                "order": 0,
                "default": "OAuth2.0"
            },
            "client_id": {
                "type": "string",
                "title": "Client ID",
                "description": "Enter your application's Client ID",
                "airbyte_secret": true
            },
            "access_token": {
                "type": "string",
                "title": "Access Token",
                "description": "Enter you application's Access Token",
                "airbyte_secret": true
            },
            "client_secret": {
                "type": "string",
                "title": "Client Secret",
                "description": "Enter your application's Client secret",
                "airbyte_secret": true
            },
            "refresh_token": {
                "type": "string",
                "title": "Refresh Token",
                "description": "Enter your application's Refresh Token",
                "airbyte_secret": true
            }
        }
    },
    {
        "type": "object",
        "order": 1,
        "title": "Key Pair Authentication",
        "required": [
            "private_key"
        ],
        "properties": {
            "auth_type": {
                "enum": [
                    "Key Pair Authentication"
                ],
                "type": "string",
                "const": "Key Pair Authentication",
                "order": 0,
                "default": "Key Pair Authentication"
            },
            "private_key": {
                "type": "string",
                "title": "Private Key",
                "multiline": true,
                "description": "RSA Private key to use for Snowflake connection. See the <a href=\"https://docs.airbyte.com/integrations/destinations/snowflake\">docs</a> for more information on how to obtain this key.",
                "airbyte_secret": true
            },
            "private_key_password": {
                "type": "string",
                "title": "Passphrase",
                "description": "Passphrase for private key",
                "airbyte_secret": true
            }
        }
    },
    {
        "type": "object",
        "order": 2,
        "title": "Username and Password",
        "required": [
            "password"
        ],
        "properties": {
            "password": {
                "type": "string",
                "order": 1,
                "title": "Password",
                "description": "Enter the password associated with the username.",
                "airbyte_secret": true
            }
        }
    }
]
```

I can see two ways of fixing this:
* Update the connector to include the auth_type property for the password case as well before merging this PR. However, without a migration it would preselect the wrong condition in the form for existing configured destinations (no data loss would occur) - not sure whether this would be OK
* Adjust the frontend logic to handle the case by allowing one condition to _not_ have a const key (by exclusion that would be enough to make it work) - however that seems hacky and ad hoc.

I'm going to prepare the PR for updating the connector spec because we should definitely do this no matter which direction we decide to take.

Interested in your thoughts @lmossman

File: airbyte-webapp/src/components/common/ApiErrorBoundary/ApiErrorBoundary.tsx
Comment (lmossman) [inline airbyte-webapp/src/components/common/ApiErrorBoundary/ApiErrorBoundary.tsx:85]: It would be nice if this could include a second argument to add some metadata about these errors, so they are easier to discover in datadog. I'm thinking of something like how errors are tracked in the useAuthenticator hook: https://github.com/airbytehq/airbyte/blob/2392acb845891f9c031b06feee9af2a2f975f1a7/airbyte-webapp/src/views/Connector/ConnectorForm/useAuthentication.tsx#L156-L159

If this could include an `id` like "formBuildError",  and also attach the connector ID to it, then we could see which connector the error occurred for. This seems important because I don't know how we would address a generic form build error that we see in datadog unless we know what connector it occurred for.

**Review by lmossman — COMMENTED** [review]
Just had one comment about improving error tracking a bit, but I don't think it is blocking.

I think the approach you laid out for dealing with the Snowflake destination issue makes sense, though it does raise one concern: 

If OSS users upgrade their airbyte version to pull in these webapp changes, but they don't update their snowflake destination connector version to the one with your fix, then they will see the `Spec uses oneOf without a shared const property` error message if they try to create or edit any snowflake connector.

This may not be a huge concern as they would just need to update that connector in order to fix it -- we just need to be sure to communicate that clearly in the release notes. But is it worth exploring a way for us to handle that case without completely failing the webapp?

**Comment by flash1293** [comment]
> If OSS users upgrade their airbyte version to pull in these webapp changes, but they don't update their snowflake destination connector version to the one with your fix, then they will see the Spec uses oneOf without a shared const property error message if they try to create or edit any snowflake connector

agreed, this needs to be made obvious.

> But is it worth exploring a way for us to handle that case without completely failing the webapp?

I honestly can’t think of a good way to do so without building a whole separate system of tracking the state which kind of goes counter the idea of the refactoring. Do you have an idea?

**Comment by lmossman** [comment]
@flash1293 no I think I agree that it would just add more complexity to track the state in a way that we don't even want to support, so I don't think it makes sense. I think your idea of adding something like 
> Please ensure your connector is updated to the latest version

is a good approach for making that clear.

**Comment by flash1293** [comment]
> to the error message is a good approach for making that clear, so let's go with that

done

> It would be nice if this could include a second argument to add some metadata about these errors, so they are easier to discover in datadog

done, also refactored the hooks a bit as it made things easier - now the connector form only knows a single "useBuildForm" hook which does all the things.

File: airbyte-webapp/src/components/common/ApiErrorBoundary/ApiErrorBoundary.tsx
Comment (lmossman) [inline airbyte-webapp/src/components/common/ApiErrorBoundary/ApiErrorBoundary.tsx:87]:
```suggestion
        connectorDefinitionId: error.connectorId,
```
nit suggestion to make it clear that this is a connector definition ID (since connector ID usually refers to the ID of an instantiated source or destination record). 

Could you also update the field on the FormBuildError class to be `connectorDefinitionId` instead of `connectorId`?

**Review by lmossman — APPROVED** [review]
Just had one more small suggestion, otherwise this LGTM and I didnt see any unexpected behavior during testing.

Reminder: this should not be merged until https://github.com/airbytehq/airbyte/pull/20566 is published and merged

**Pull request merged**
